app/models/challenge_signup_summary.rb

```diff
@@ -38,6 +51,85 @@ def tag_summary(tag)
     end
   end
 
+  public
+
+  ######################################################################
+  # GENERATING THE SUMMARY IN RESQUE
+  ######################################################################
+
+  @queue = :collection
+
+  # The action to be performed when this class is enqueued using Resque.
+  def self.perform(collection_id)
+    collection = Collection.find(collection_id)
+    summary = ChallengeSignupSummary.new(collection)
+    summary.generate_in_background
+  end
+
+  # Asynchronously generate the cached version.
+  def enqueue_for_generation
+    Resque.enqueue(ChallengeSignupSummary, collection.id)
+  end
+
+  # Write the summary to Memcached, so that it can be retrieved and displayed.
+  def generate_in_background
+    locals = {
+      challenge_collection: collection,
+      tag_type: self.tag_type,
+      summary_tags: self.summary,
+      generated_live: false
+    }
+
+    partial = "challenge/#{self.challenge.class.name.demodulize.tableize.singularize}/challenge_signups_summary"
+
+    self.cached_contents = ChallengeSignupsController.render(partial: partial,
+                                                             locals: locals)
+  end
+
+  ######################################################################
+  # CACHING THE SUMMARY
+  ######################################################################
+
+  # Retrieve the value of the cache for this signup summary.
+  def cached_contents
+    cached_info[:contents]
+  end
+
+  # Retrieve the time that the cache was last updated.
+  def cached_time
+    cached_info[:time]
+  end
+
+  # The equivalent of touching a file in the file system. Update the
+  # modification time.
+  def touch_cache
+    data = cached_info.dup
+    data[:time] = Time.now
+    Rails.cache.write(cache_key, data)
```

Comment:
I tried to mimic the old behavior as much as possible. While the old code did set up the summary to be regenerated after an hour, it also deliberately kept the old summary around after that hour had elapsed, so that the old summary could be displayed even while the new summary was being generated in the background. (And it also kept the old sign-up summary indefinitely once sign-ups were closed.) The new code is intended to work in much the same way.

The disadvantage of setting `expires_in` is that it deletes the old sign-up summary, resulting in a period where the old sign-up summary is missing, but the new sign-up summary hasn't been generated yet. That's not necessarily a problem, particularly if generating the summary is snappy and the Resque queue is short, but it's also not ideal. IMHO, if `expires_in` is set at all, it should be set to something fairly long.

(If the summary should be regenerated when the challenge is closed -- which is one of the things that setting `expires_in` would accomplish -- it would probably be simplest to add another clause to the timeout check in `ChallengeSignupsController` instead: e.g. `modification_time < 24.hours.ago`. That way, there'll always be a sign-up summary to display, even if it's a little out of date. Well, unless Memcached runs out of room and starts deleting old keys.)